xds: fix various bugs and simplify control flow

marseel opened the pull request:
Please review commit by commit, it should be much simpler. 

High level summary:
- Added two tests that shows that current implementation had bugs - TestTheSameVersionOnRestart and TestNotAckedAfterRestart:
1. TestTheSameVersionOnRestart shows that we were not sending response in case of new stream while accidentally previous versionInfo was equal to current version in xds cache.
2. TestNotAckedAfterRestart shows that in case of reciving NACK on a new stream, we were still ACKing observers instead of NACKs.
- Fixed two previous tests that were just incorrect - TestRequestStaleNonce and TestRequestHighVersionFromTheStart - fixing them did not break previous implementation. 
1. Test TestRequestStaleNonce previously used nonce=0 and versionInfo=2. We had special handling for nonce=0. if we only switched to nonce=1 an versionInfo=2 it would result in hitting `ErrInvalidResponseNonce` case. So instead switch to both versionInfo and nonce being stale. 
2. TestRequestHighVersionFromTheStart - just had nonce set to versionInfo, but first request on stream needs nonce empty.
Following commits after tests are simplifying control plane and fixing control plane for newly added tests.


```release-note
xds: Fix a case in which after cilium-agent we were not sending updated resources to Envoy
```

marseel committed:
xds: simplify handling of NACKs.

Previously, we were overriding versionInfo with most recent versionInfo
that was send to client. However, in case of stale nonce received, we
want to resend response based on stale nonce. This happens only when
stale request was NACK.

Signed-off-by: Marcel Zieba <marcel.zieba@isovalent.com>

marseel committed:
xds: simplify indentation

No function changes, just one less indent.

Signed-off-by: Marcel Zieba <marcel.zieba@isovalent.com>

marseel committed:
xds: do not ACK observers on first request

Start ACKing observers only if we have sent a response that was received
by client. However, this does not fully address test
TestNotAckedAfterRestart in case first response is NACK.

Signed-off-by: Marcel Zieba <marcel.zieba@isovalent.com>

marseel committed:
xds: simplify watching resource changes

We should not wait based on ACKed version by client, but rather than
that, we should wait for next version after version that was NACKed.

Signed-off-by: Marcel Zieba <marcel.zieba@isovalent.com>

marseel committed:
xds: do not override nonce anymore

This has an effect of starting with nonce=0 on first request in stream.
As a result, WatchResources will receive latestVersion=0 and return
response immadietly.

Signed-off-by: Marcel Zieba <marcel.zieba@isovalent.com>

marseel committed:
xds: handle ACKs/NACKs correctly for new stream

This fixes test TestNotAckedAfterRestart. Previously if first response
on a new stream was NACK, we were still ACKing observers.

Signed-off-by: Marcel Zieba <marcel.zieba@isovalent.com>

marseel committed:
xds: fix test for handling stale nonce

Previously, we had a special handling for nonce=0 in xds server.
However, this test case is meant to handle stale nonce, not necessarily
nonce=0. Bump nonce to 1 to show that current implementation does not
handle stale nonce correctly.
Additionally, I don't think it makes sense to send a new versionInfo with
stale nonce as we did previously.
Alternative approach would be to not send response at all for stale nonce.

Signed-off-by: Marcel Zieba <marcel.zieba@isovalent.com>

marseel committed:
xds: add test that shows broken behaviour on restart

In some cases, it is possible that client opens a new stream with the
same version as the one stored in xds cache. In this case, we should
send full response back, but this was not the case. With current
implementation this test fails with an error:
level=error msg="error while receiving request from xDS stream"
error="context deadline exceeded"
on line server_e2e_test.go:1169

Signed-off-by: Marcel Zieba <marcel.zieba@isovalent.com>

marseel committed:
xds: Fix bumping version on restart

Previously, we were bumping resource version only if it was smaller than
current version. However in case that previous version was equal to new
version after restart, we were not sending response. This fixes test
TestTheSameVersionOnRestart

Signed-off-by: Marcel Zieba <marcel.zieba@isovalent.com>

marseel committed:
xds: add test to test phantom ACKs

In the case of the agent restart and reciving request with versionInfo
higher than current one, if response is NACKed by client, with current
implementation we issue ACKs instead of NACKs. This test covers this
case and current implementation fails it.

Signed-off-by: Marcel Zieba <marcel.zieba@isovalent.com>

marseel committed:
xds: fix test TestRequestHighVersionFromTheStart

First request on new stream always has empty nonce.

Signed-off-by: Marcel Zieba <marcel.zieba@isovalent.com>

jrajahalme commented on pkg/envoy/xds/server.go:
```diff
@@ -194,11 +194,6 @@ type perTypeStreamState struct {
-	// if a new request is an ACK (VersionInfo matches current version), or a NACK
-	// (VersionInfo matches an earlier version).
-	version uint64
-
```
We had to use this field before response nonce existed to detect ACK vs. NACK. Now that we set response nonce to be the same as the version and the nonce is reflected back in the next request, we don't strictly need this field any more.

jrajahalme commented on pkg/envoy/xds/server.go:
```diff
@@ -395,7 +391,7 @@ func (s *Server) processRequestStream(ctx context.Context, streamLog *logrus.Ent
 
 			requestLog.Debugf("starting watch on %d resources", len(req.GetResourceNames()))
-			go watcher.WatchResources(ctx, typeURL, versionInfo, nodeIP, req.GetResourceNames(), respCh)
+			go watcher.WatchResources(ctx, typeURL, nonce, nodeIP, req.GetResourceNames(), respCh)
```
`nonce` here is the last version we sent, regardless if it was ACKed of NACKed by the client. Maybe we should rename this local variable to `lastVersionSent` or something? That way it would be clear that we are waiting for a version bump to happen regardless if the last version was ACKed or NACKed?

marseel replied:
> nonce here is the last version we sent, regardless if it was ACKed of NACKed by the client.

This is not fully true. `nonce` is the last version that **client saw and reported back to us**. In theory, it doesn't necessarily need to be last version that we sent. I was on a verge if we should return response based on this or what we had previously as `version uint64` that actually indicated last version that we sent. I think sending back response based on this `nonce` is safer than previously used `version uint64`. Worst case, we would send response one additional time in case of "lagging" client reporting stale `nonce`.

How about name like `lastReceivedVersion` instead?

jrajahalme commented on pkg/envoy/xds/server.go:
```diff
@@ -351,12 +351,12 @@ func (s *Server) processRequestStream(ctx context.Context, streamLog *logrus.Ent
 			}
-			if nonce == 0 && versionInfo > 0 {
-				requestLog.Infof("xDS was restarted, setting nonce to %d", versionInfo)
-				nonce = versionInfo
```
Did you consider that if we send the nonce as `0` in the first response, we'll also receive it back as `0` on the following request?

marseel replied:
Good point, let me add the test and ensure that we are handling it correctly. Although, right now we would send versionInfo=0 on first request only if there was nothing in local cache, as each update bumps local version.

marseel replied:
hah, so It seems like we are never sending first response with nonce=0.
That was implemented in https://github.com/cilium/cilium/pull/4535/commits/374c83671c68721e165a479e8d4b1b55d3377a4a
as part of simplifying of not having pointer and just uints.
Specifically, line: https://github.com/cilium/cilium/blob/d4e01a5cb737b8710c9e347663a9e05ef24777c1/pkg/envoy/xds/watcher.go#L46
ensure that we will always return at least nonce=1.

jrajahalme replied:
Does this affect the logic/fixes in this PR?

marseel replied:
> Does this affect the logic/fixes in this PR?

No, it just means that this case:
> Did you consider that if we send the nonce as 0 in the first response, we'll also receive it back as 0 on the following request?

will never happen. Either nonce is zero and we haven't send anything to client or nonce is no-zero and we send something to client + client responded with it. Essentially meaning that `clientReceivedFirstResponse` is correct.

jrajahalme commented on pkg/envoy/xds/server.go:
```diff
@@ -266,6 +266,9 @@ func (s *Server) processRequestStream(ctx context.Context, streamLog *logrus.Ent
 	clientReceivedFirstResponse := false
+	// firstResponseAcked indicates if we already
+	// had some request on this stream that was ACKed by a client.
+	firstResponseAcked := false
```
Based on the comment it does not need to be *the first* response that was ACKed, but any response. If so, maybe rename as `responseAcked` or `aResponseAcked`, or `haveAckedResponse`?

marseel replied:
Make sense

marseel replied:
Switched to `responseAcked`

jrajahalme reviewed (commented):
It looks like we use the first request to sync the versions between the client and the server, and we always send a response on the first request. We also want to get an ACK/NACK to the first response to complete that sync.

Could we use this framing to clarify the code/commentary even further?

marseel committed:
xds: rename versionInfo/nonce

Let's use more descriptive names like lastAppliedVersion and
lastReceivedVersion.

Signed-off-by: Marcel Zieba <marcel.zieba@isovalent.com>

marseel reviewed (commented):
Pushed some changes to versionInfo/nonce in last commit to make it more descriptive + renamed `firstResponseAcked` to `responseAcked`
